File: src/main/services/history.ts
/**
 * Per-project chat history persistence so a conversation survives app restarts.
 *
 * The Copilot CLI session id is persisted on the project (see store.ts), which
 * preserves the *agent's* memory; this module persists the *UI transcript* so
 * the two stay in sync. Each project's messages are stored in their own JSON
 * file under `userData/chats/` to avoid rewriting the whole app state (and to
 * keep large transcripts out of the main store file).
 */

import { app } from 'electron'
import { join } from 'path'
import { existsSync, mkdirSync, readFileSync, writeFileSync, rmSync } from 'fs'
import type { ChatMessage } from '../../shared/ipc'
import { MAIN_THREAD_ID } from '../../shared/ipc'

/** Keep transcripts bounded; older messages beyond this are dropped on save. */
const MAX_MESSAGES = 1000

function chatsDir(): string {
  return join(app.getPath('userData'), 'chats')
}

/**
 * Restrict the on-disk filename to a safe slug derived from the project id (and
 * thread id, for side threads). The main thread keeps the bare `<projectId>.json`
 * name for backward compatibility with transcripts written before side threads.
 */
function historyFile(projectId: string, threadId: string = MAIN_THREAD_ID): string {
  const safeProject = projectId.replace(/[^a-zA-Z0-9_-]/g, '').slice(0, 128) || 'unknown'
  if (threadId === MAIN_THREAD_ID) return join(chatsDir(), `${safeProject}.json`)
  const safeThread = threadId.replace(/[^a-zA-Z0-9_-]/g, '').slice(0, 128) || 'thread'
  return join(chatsDir(), `${safeProject}__${safeThread}.json`)
}

/** Coerce arbitrary persisted JSON into a clean ChatMessage[]. */
function sanitize(input: unknown): ChatMessage[] {
  if (!Array.isArray(input)) return []
  const out: ChatMessage[] = []
  for (const item of input) {
    const m = item as Partial<ChatMessage>
    if (!m || (m.role !== 'user' && m.role !== 'assistant')) continue
    out.push({
      id: String(m.id ?? ''),
      role: m.role,
      text: typeof m.text === 'string' ? m.text : '',
      tools: Array.isArray(m.tools) ? m.tools : [],
      error: typeof m.error === 'string' ? m.error : undefined,
      attachments: typeof m.attachments === 'number' ? m.attachments : undefined
    })
  }
  return out
}

/** Load a project thread's persisted conversation (empty array when none/invalid). */
export function loadHistory(projectId: string, threadId?: string): ChatMessage[] {
  try {
    return sanitize(JSON.parse(readFileSync(historyFile(projectId, threadId), 'utf8')))
  } catch {
    return []
  }
}

/** Persist a project thread's conversation. An empty array removes the file. */
export function saveHistory(projectId: string, messages: ChatMessage[], threadId?: string): void {
  const clean = sanitize(messages).slice(-MAX_MESSAGES)
  const file = historyFile(projectId, threadId)
  if (clean.length === 0) {
    try {
      rmSync(file, { force: true })
    } catch {
      /* best-effort */
    }
    return
  }
  const dir = chatsDir()
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true })
  writeFileSync(file, JSON.stringify(clean), 'utf8')
}

/** Delete a project thread's persisted conversation (used on removal). */
export function clearHistory(projectId: string, threadId?: string): void {
  try {
    rmSync(historyFile(projectId, threadId), { force: true })
  } catch {
    /* best-effort */
  }
}

File: src-tauri/src/services/history.rs
//! Per-project chat-history persistence — the Rust counterpart to
//! `src/main/services/history.ts`. Each project's transcript is a JSON file
//! `<projectId>.json` under `<dataDir>/chats/`.

use super::paths;
use crate::types::ChatMessage;

/// Keep transcripts bounded; older messages beyond this are dropped on save.
const MAX_MESSAGES: usize = 1000;

fn safe_slug(input: &str, fallback: &str) -> String {
  let s: String = input
    .chars()
    .filter(|c| c.is_ascii_alphanumeric() || *c == '_' || *c == '-')
    .take(128)
    .collect();
  if s.is_empty() {
    fallback.to_string()
  } else {
    s
  }
}

fn history_file(project_id: &str) -> std::path::PathBuf {
  paths::chats_dir().join(format!("{}.json", safe_slug(project_id, "unknown")))
}

/// Coerce arbitrary persisted JSON into a clean `Vec<ChatMessage>`, dropping
/// legacy "merge" system events from the removed side-threads feature.
fn sanitize(messages: Vec<ChatMessage>) -> Vec<ChatMessage> {
  messages
    .into_iter()
    .filter(|m| (m.role == "user" || m.role == "assistant") && m.kind.as_deref() != Some("merge"))
    .collect()
}

/// Load a project's persisted conversation (empty when none/invalid).
pub fn load_history(project_id: &str) -> Vec<ChatMessage> {
  match std::fs::read_to_string(history_file(project_id)) {
    Ok(raw) => serde_json::from_str::<Vec<ChatMessage>>(&raw)
      .map(sanitize)
      .unwrap_or_default(),
    Err(_) => vec![],
  }
}

/// Persist a project's conversation. An empty list removes the file.
pub fn save_history(project_id: &str, messages: Vec<ChatMessage>) {
  let mut clean = sanitize(messages);
  if clean.len() > MAX_MESSAGES {
    clean = clean.split_off(clean.len() - MAX_MESSAGES);
  }
  let file = history_file(project_id);
  if clean.is_empty() {
    let _ = std::fs::remove_file(&file);
    return;
  }
  if let Err(e) = std::fs::create_dir_all(paths::chats_dir()) {
    log::warn!("failed to create chats dir {}: {e}", paths::chats_dir().display());
  }
  if let Ok(text) = serde_json::to_string(&clean) {
    if let Err(e) = std::fs::write(&file, text) {
      log::warn!("failed to save chat history to {}: {e}", file.display());
    }
  }
}

/// Delete a project's persisted conversation (used on removal).
pub fn clear_history(project_id: &str) {
  let _ = std::fs::remove_file(history_file(project_id));
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn design_apply_markers_survive_chat_history_sanitizing_and_serde() {
    let rows: Vec<ChatMessage> = serde_json::from_value(serde_json::json!([
      {"id":"apply-user","role":"user","text":"Apply the visual draft","designApplyId":"apply-1"},
      {"id":"apply-1","role":"assistant","text":"","designApplyId":"apply-1","interrupted":true},
      {"id":"old-merge","role":"assistant","kind":"merge","text":"Legacy merge"}
    ])).unwrap();
    let saved = serde_json::to_string(&sanitize(rows)).unwrap();
    let reloaded = sanitize(serde_json::from_str(&saved).unwrap());
    assert_eq!(reloaded.len(), 2);
    assert!(reloaded.iter().all(|row| row.design_apply_id.as_deref() == Some("apply-1")));
    assert_eq!(reloaded[1].interrupted, Some(true));
    assert!(saved.contains("\"designApplyId\":\"apply-1\""));
    assert!(!saved.contains("design_apply_id"));
  }

  #[test]
  fn design_marker_is_optional_in_legacy_chat_history() {
    let row: ChatMessage = serde_json::from_str(r#"{"id":"ordinary","role":"user","text":"Hello"}"#).unwrap();
    assert!(row.design_apply_id.is_none());
    assert!(!serde_json::to_string(&row).unwrap().contains("designApplyId"));
  }
}
